[feature] Added Cluster Prevention Mechanism #171

cestercian opened the pull request:
## Checklist

- [x] I have read the [OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html).
- [x] I have manually tested the changes proposed in this pull request.
- [x] I have written new test cases for new code and/or updated existing tests for changes to existing code.
- [x] I have updated the documentation.

## Reference to Existing Issue

Closes #171

## Description of Changes

This PR addresses issue #171 by preventing the overlapping of clusters in the `netjsongraph.js` visualization.  

### Changes Implemented:
- Introduced spatial indexing using `KDBush` to optimize clustering performance.
- Implemented grouping of nodes based on proximity and clustering attributes.
- Modified cluster assignment logic to prevent overlapping.
- Updated centroid calculations for accurate cluster placement.
- Ensured better handling of non-clustered nodes.

### Testing  
A test file (`cluster_overlap_test.html`) was created to visualize and verify the clustering behavior. The test file simulates overlapping nodes with different statuses and ensures they are properly clustered without overlaps.

## Screenshot
<img width="1497" alt="image" src="https://github.com/user-attachments/assets/025ce6e3-28c5-4ff3-941e-80c2684146f9" />

nemesifier reviewed (commented):
This has slipped my attention, the notification for this was buried among many other notifications. That's why it's good to let us know about your work in the [dev chat](https://matrix.to/#/#openwisp_development:gitter.im).

nemesifier reviewed (commented):
In the PR description you mention a test file called `cluster_overlap_test.html`, which I see you have removed. What type of problem where you having with this? I think such a test file to compare the behavior of this branch with master and verify whether the situation really improves is quite useful.

cestercian commented:
Thanks for the review @nemesifier! The cluster_overlap_test.html file was initially created to visualize the issue, but I removed it since I thought it wasn't necessary after fixing the overlap problem. However, I see your point—it could be useful for verifying improvements. I can reintroduce it and ensure it effectively compares the branch behavior with master. Let me know if you'd like any specific tests included!

nemesifier commented on examples/netjson-cluster-overlap.html:
```diff
@@ -0,0 +1,153 @@
+<!DOCTYPE html>
```
This file is in `/examples/netjson-cluster-overlap.html`, but most examples are stored in `/public/example_templates`, why is that? Can you please:

- Make this example consistent with the other examples
- Add a link to this example in the index?

If all goes well we could replace the current clustering example with this one as it's more complete.

cestercian commented:
Hi @nemesifier ,

I've made the requested changes:

✅ Moved netjson-cluster-overlap.html to /public/example_templates for consistency.
✅ Updated index.html to include a link to the new example, styled like the others.

Let me know if you need any further modifications or if you'd like me to proceed with replacing the current clustering example with this one. Thanks for your feedback!

nemesifier reviewed (commented):
This is what I see when I try to open the example, do you see a different situation?

`Cannot GET /example_templates/netjson-cluster-overlap.html`.

![Screenshot from 2025-03-19 19-01-01](https://github.com/user-attachments/assets/3907278a-fb05-4655-8824-2c561e3dd90d)

cestercian commented:
hey @nemesifier 

I’ve fixed the issue. The error occurred because the requested file was not being served correctly. I’ve updated `netjson-cluster-overlap.html` and changed the file location in `index.html` my latest commit, which should resolve it.

Let me know if you still encounter any issues!

nemesifier reviewed (commented):
I don't know if I am doing something wrong but this isn't working for me, here's what I see:

[netjsongraphjs-clustering-pr.webm](https://github.com/user-attachments/assets/fd25dd21-133f-4b3e-a784-169901a1608c)

@niteshsinha17 do you see a different situation or do you get the same result I get?

cestercian commented:
<img width="975" alt="image" src="https://github.com/user-attachments/assets/e9cd4ec6-16c9-4e2e-bc8f-c7ea352db545" />

It looks like the `NetJSONGraph` library isn't being recognized. This often happens when the project isn't started properly.  

@nemesifier, are you running `yarn start`, or are you just opening the `index.html` file directly? If you're just opening the file, the dependencies might not be loading correctly. Try starting the project with `yarn start` and let me know if the issue persists.

niteshsinha17 commented:
> I don't know if I am doing something wrong but this isn't working for me, here's what I see:
> 
>  netjsongraphjs-clustering-pr.webm 
> @niteshsinha17 do you see a different situation or do you get the same result I get?

@nemesifier I am getting this result. Seems something is wrong. I will look into PR in detail.

![Screenshot 2025-03-22 at 8 58 06 PM](https://github.com/user-attachments/assets/a12e36e9-c6f6-486e-ac54-9766487eb235)

niteshsinha17 reviewed (changes requested):
I have done a very basic review for now. The graph is not rendering properly. 

Check this comment, here @nemesifier 
 has mentioned about an example that needs to be added. https://github.com/openwisp/netjsongraph.js/pull/239#pullrequestreview-1899551939

cestercian commented on examples/netjson-cluster-overlap.html:
```diff
@@ -0,0 +1,153 @@
+    </style>
+  </head>
+  <body>
+    <script src="../dist/netjsongraph.min.js"></script>
```
Without this line, the browser has no access to the NetJSONGraph library code, meaning:

The new NetJSONGraph() constructor would be undefined
All map rendering would fail
Clustering logic wouldn't execute

niteshsinha17 replied:
I think, it's not required. check other examples.

cestercian replied:
Reply to @niteshsinha17:
I removed the line `<script src="../dist/netjsongraph.min.js"></script>`, and the map still rendered correctly. It looks like NetJSONGraph is already being loaded elsewhere in the project. Thanks for pointing it out! 😁

cestercian commented on examples/netjson-cluster-overlap.html:
```diff
@@ -0,0 +1,153 @@
+        background: rgba(255, 255, 255, 0.8);
+        padding: 10px;
+        border-radius: 4px;
+      }
```
update the style in the recent commit

cestercian commented on examples/netjson-cluster-overlap.html:
```diff
@@ -0,0 +1,153 @@
+    <meta name="description" content="NetJSON Cluster Overlap Example" />
+    <meta name="viewport" content="width=device-width, initial-scale=1" />
+    <link rel="stylesheet" href="../src/css/netjsongraph.css" />
+    <link rel="stylesheet" href="../src/css/netjsongraph-theme.css" />
```
updated the src path in the recent commit

cestercian commented:
Thank you, @niteshsinha17, for your review and valuable feedback!  

I've made the following updates based on your suggestions:  

- Updated the CSS import paths to ensure they are loaded from the library, keeping it consistent with other examples.  
- Adjusted the styling of the legend to align with existing examples for a uniform look.  
- Retained the `<script src="../dist/netjsongraph.min.js"></script>` line, as it is essential for loading the NetJSONGraph library.  
  - Without it, the `new NetJSONGraph()` constructor would be undefined, causing the graph rendering and clustering logic to fail.  

Let me know if there's anything else you'd like me to refine. Thanks again for your insights!

and FYI 😊 the changes that reflect in projects are made to file in `example templates` directory I don't know why this issue persists Im trying to find the bug, but the changes in the file `examples/netjson-cluster-overlap.html` doesn't reflect in the demo

cestercian commented:
<img width="1495" alt="image" src="https://github.com/user-attachments/assets/14cd54a9-0d86-44cf-b239-0869c12db86d" />

this is output I'm getting, let me know if there are any other issues, Thanks again for the review !!

nemesifier reviewed (changes requested):
Now I can see the example, thank you for fixing that!

Can you please create more nodes (eg: 20-30) so that we end up seeing 1 cluster for each status when opening up the example?

Can you also make sure that when zooming in, we can see the circle of the same color of its status? Right now it shows as blue, but in the OpenWISP Dashboard each circle is of the color of its status, see the following screenshot from the [OpenWISP Demo](https://openwisp.org/demo/):

![Screenshot from 2025-03-24 18-20-37](https://github.com/user-attachments/assets/28f9e458-65d9-4799-8acc-fca59bc8ec94)

We need to improve the clustering feature to avoid the overlap because on bigger instances with many nodes this is how the map shows up:

![Screenshot from 2025-03-24 18-24-53](https://github.com/user-attachments/assets/bcb29743-fd8b-4e8e-b113-e110093117a5)

cestercian commented:
I've completed the first task. However, for the second task, implementing this change requires modifying `src/js/netjsongraph.config.js`, which would affect the map rendering across the entire project. 

Since this PR is already quite long, I think it would be best to handle this in a separate PR. Would you be open to creating a new issue for this, so we can track and work on it separately?

cestercian commented on examples/netjson-cluster-overlap.html:
>In the PR description you mention a test file called `cluster_overlap_test.html`, which I see you have removed. What type of problem where you having with this? I think such a test file to compare the behavior of this branch with master and verify whether the situation really improves is quite useful.

_Originally posted by @nemesifier in https://github.com/openwisp/netjsongraph.js/pull/349#pullrequestreview-2700074789_

cestercian commented:
Moved cluster overlap prevention logic to `lib/js/clusterUtils.js` for better modularity.

## Changes Made
- Extracted `preventClusterOverlap()` into `clusterUtils.js`
- Added `setupClusterOverlapPrevention()` for centralized event handling
- Updated `netjson-cluster-overlap.html` to use ES module imports
- Improved `README.md` with project structure and usage details

nemesifier reviewed (changes requested):
This is what I am seeing:

[Peek 2025-03-27 16-56.webm](https://github.com/user-attachments/assets/af18828f-081a-4d77-95ff-6d837fd77777)

When zooming in, the circles should appear of the color of their status, not blue.
When zooming out, I expect to see a few clusters of different colors, but I see they're overlapping.

Please, in the next iteration post a similar GIF/Webp recording of the result, I am using a Linux program called Peek which is very handy.

cestercian commented:
> This is what I am seeing:
> 
>  Peek.2025-03-27.16-56.webm 
> 
> When zooming in, the circles should appear of the color of their status, not blue. When zooming out, I expect to see a few clusters of different colors, but I see they're overlapping.
> 
> Please, in the next iteration post a similar GIF/Webp recording of the result, I am using a Linux program called Peek which is very handy.

Thanks for the feedback, @nemesifier !

I already mentioned the need to modify nodeStyle.color in src/js/netjsongraph.config.js (line 172) in my earlier comment. Since this change affects the entire project’s map rendering, I suggested handling it in a separate PR, since this PR is getting to long already.

Would you be open to creating a new issue for this so we can track it separately?

cestercian commented:
<img width="1492" alt="image" src="https://github.com/user-attachments/assets/9ec5bb5c-f447-478c-bbc4-87853eac4040" />

when Zoomed IN

<img width="1492" alt="image" src="https://github.com/user-attachments/assets/008c13cd-c2be-4388-9d2c-9600c0f126fa" />

When Zoomed OUT

I've made the necessary changes to improve the cluster separation and have included updated images for review. Let me know if there's anything else you'd like to tweak!

cestercian commented:
Hey @nemesifier is this what were you talking about 
<img width="1495" alt="image" src="https://github.com/user-attachments/assets/4a1015f3-49a8-46b7-a4f3-f5f8b2837954" />
when zoomed out

cestercian commented:
https://github.com/user-attachments/assets/ce5774ea-0e0a-42ec-b305-dab09f9fa55b

nemesifier reviewed (changes requested):
It's looking a lot better now! Promising! :clap: 

![Screenshot from 2025-04-10 20-08-20](https://github.com/user-attachments/assets/485609ee-72a1-4c30-bb49-f2836d6234ac)

I haven't looked at the code much yet, but I will do that once the example looks good, I think there's still a few details to improve:

1. You can now replace the main cluster example with your example
2. Please vary the number of nodes in each cluster so that we do not have 10 nodes in each, we could have 22, 15, 7, any number is fine but let's vary this
3. When zooming in and the cluster expands, the nodes are blue, that's not right, the nodes should be of the color of their status, is this clear? If not I'll send a screenshot but I hope my text description is good enough for now, let me know.

Once you're done with the changes, please record a GIF or a video of the result and post it in the reply to make the reviewer's work easier.

cestercian commented:
https://github.com/user-attachments/assets/bdc14039-3e95-4360-99f1-61cc264db0f0


Hey @nemesifier , @niteshsinha17 ! 
I've made a couple of the updates:

- Replaced the main cluster example with my own.
- Adjusted the number of nodes in each cluster so they now vary (e.g. 22, 15, 7, etc.).

However, I haven't yet implemented the fix for the node colors when zooming in — currently, they still appear blue instead of reflecting their actual status color, currently working on it.

Let me know if anything still looks off.

cestercian commented:
https://github.com/user-attachments/assets/d37bb25e-8821-488c-9413-06196f706474

Applied Node Coloring based on node status

nemesifier commented on lib/js/clusterUtils.js:
```diff
@@ -0,0 +1,76 @@
+    
+    leafletMap.on('moveend', preventClusterOverlap);
+    
+    
```
why so much whitespace? Let's avoid this.

nemesifier commented on lib/js/clusterUtils.js:
```diff
@@ -0,0 +1,76 @@
+    
+    console.warn('[NetJSONGraph] setupClusterOverlapPrevention: Leaflet map instance is required for cluster overlap prevention.');
+  }
+} 
```
Ensure all files always end with blank lines.

nemesifier reviewed (changes requested):
Looks good @cestercian! The test coverage has went down considerably, can you please fix that? Look at the CI build configuration to see how to calculate test coverage locally so you can see which lines are not being executed during tests.

See also my comments below. Tomorrow I will do more code review, now I only had time to do a quick test.

cestercian commented:
Hey @niteshsinha17 ,
I moved overlap prevention logic from lib/js/clusterUtils.js to src/js/cluster-utils.js, updated imports, and removed the old file.
and optimized getNodeStyle() usage in netjsongraph.render.js by properly using all returned style configs.

let me know if any other changes are needed 😁

nemesifier reviewed (commented):
I would group all the clustering docs into a single section coming just before "Example Usage". The rest looks good to me but I'm trying to spare some time to do a complete round of testing before merging.

niteshsinha17 commented on src/js/netjsongraph.render.js:
```diff
@@ -267,15 +267,74 @@ class NetJSONGraphRender {
 
     const series = [
-      Object.assign(configs.mapOptions.nodeConfig, {
+      {
         type:
           configs.mapOptions.nodeConfig.type === "effectScatter"
             ? "effectScatter"
             : "scatter",
+        name: "nodes",
         coordinateSystem: "leaflet",
         data: nodesData,
         animationDuration: 1000,
-      }),
+        label: configs.mapOptions.nodeConfig.label,
```
@cestercian  By looking at your code and spending some time in debugging. I think now I know the reason why u are adding an extra object in the `series` array. You have added this because `getNodeStyle` is returning {} as size for some nodes (may be because of config) and you wanted to provide your custom functions so that nodes get correct styles?

nemesifier reviewed (changes requested):
While testing https://github.com/openwisp/openwisp-monitoring/pull/668 on demo I see that there's overlapping clusters:

[demo.webm](https://github.com/user-attachments/assets/a6120027-b66e-46c0-9c26-61ce730db8aa)

Will DM you to show you this.

nemesifier reviewed (commented):
[denmark.webm](https://github.com/user-attachments/assets/2dc39f5d-c2a7-4beb-aacd-555d3d5e797b)

cestercian commented:
https://github.com/user-attachments/assets/8e3c6ec3-bdc3-4f15-a1ce-685f7622afee

### Logic Changes for Cluster Separation

**Enhanced Cluster Logic:**
- Implemented a mechanism to offset clusters of different attribute groups that originate from the same pixel location to prevent overlap.
- Clusters are arranged in a circular pattern  using angle-based separation (360° divided by the number of clusters) around the original position using a configurable `clusterSeparation` distance.
- The offset is calculated in pixels and then converted back to geographical coordinates using Leaflet's `containerPointToLatLng` method.

cestercian reviewed (commented):
refined and updated cluster handling with new logic still need to update the test cases for the new logic

cestercian commented:
![cluster june 6](https://github.com/user-attachments/assets/3d88fb03-11b2-4bdc-ac46-aed23ab09273)

Hey @nemesifier ,

I've updated the clustering logic to form clusters even for single nodes and added test cases to cover this change.

nemesifier reviewed (commented):
Can you please fix the following minor things?

- avoid importing the javascript file of the library twice
- move data to external JSON file (as with the other examples)
- remove these annoying node labels
  ![Screenshot from 2025-07-03 21-51-23](https://github.com/user-attachments/assets/c4f6e6dd-6efd-4a6a-8cf6-536263c6421a)

cestercian commented on src/js/netjsongraph.render.js:
```diff
@@ -267,15 +268,74 @@ class NetJSONGraphRender {
+              (configs.mapOptions.nodeConfig &&
+                configs.mapOptions.nodeConfig.nodeStyle &&
+                configs.mapOptions.nodeConfig.nodeStyle.color) ||
+              "#6c757d";
```
The node color is picked from the config if it’s defined under nodeStyle.color. If nothing’s set there, it just falls back to a default gray so that nodes always have a visible color on the map.

nemesifier reviewed (changes requested):
The API of OpenWISP Monitoring returns data in GeoJSON.

The data in this example is not GeoJSON and that's why we are having all this problems.

1. We need to make sure to test our example here looks very similar to what we'll use in OpenWISP.
2. Please start investigating why loading data in GeoJSON format triggers different rendering, we need to fix this issue as it's preventing us from testing this solution in real world scenarios and hence we are not sure the solution is really working.

cestercian commented:
> Please squash all the commits into one which will make managing both PRs at the same time easier (#396 and #171).

Done in https://github.com/openwisp/netjsongraph.js/pull/349/commits/4aaecad04a674837cacdd698caab7a90ac205f6b

nemesifier reviewed (changes requested):
During testing, I noticed this quirk:

<img width="730" height="474" alt="Screenshot from 2025-07-19 15-24-33" src="https://github.com/user-attachments/assets/c779e274-7016-4617-9bff-26492d0cb9d0" />

Can we change the code so that clusters are shown above geojson geometries?
Even better would be to consider the center of any geojson geometries as a point to include in the cluster, when the cluster breaks we show the geometry.

cestercian commented:
Hey @nemesifier I have tested the clustering PR with the changes from the streamlining PR by applying new clustering logic to the geojson sample data

cestercian commented:
Hey @nemesifier , I've addressed all the changes requested by @niteshsinha17 , and also created new test cases for the coveralls , I've also updated the clustering example to use geojson-saple.json for better testing 

Let me know if any other changes are required

cestercian commented on src/js/netjsongraph.util.js:
```diff
@@ -301,102 +304,236 @@ class NetJSONGraphUtil {
+
+    const isClusterable = (n) =>
+      !(n.properties && n.properties._featureType) ||
+      n.properties._featureType === "Point";
```
that is prettier if you are talking about indentation and if you are talking about type checking that is required for netjson and geojson data-type

nemesifier reviewed (changes requested):
When I asked to use GeoJSON as format, I didn't ask to copy the data from the geojson example to the clustering example, I asked to convert the NetJSON format to GeoJSON with the example data used in the clustering example. Can you please do this?

Please also make sure human readable lables are used, it's ok to use AI to automatically update the data, as long as the example works and the data shows as human friendly I am fine with it, with this change done we can merge.
